bullet-train-co/bullet_train-sortable
# Add sortable Stimulus controller

Resolves #4

The Stimulus-based sortable functionality this PR introduces will make it easy and resilient to update a sortable table while keeping the table sortable. This is by virtue that `dragula`, the underlying library, doesn't modify the table's elements but simply catches events at the level of the element on which it's instantiated. Also, because it's a Stimulus controller, replacing the whole table properly destroys the previous instance of dragula and also re-injects its functionality on the new elements.

## To test

1. use `bin/develop` in your main BT repo and select `bullet_train-sortable` from the list. This will create a copy of the gem in `local/bullet_train-sortable`.
2. In another Terminal window, `cd local/bullet_train-sortable`, pull this PR's branch (`add-sortable-stimulus-controller`) and wait for the `bin/develop` process in the other Terminal window to pick up the changes.
3. Within your main app, modify all instances of `data-reorder="` with `data-controller="sortable" data-sortable-reorder-path-value="`
4. Within your main app, make sure that the import for `support/jquery` is before the line to import `@rails/ujs`
5. Within your main app, add the following two lines to `app/javascript/controllers/index.js`

```js
# app/javascript/controllers/index.js

import { controllerDefinitions as sortableControllers } from "@bullet-train/bullet-train-sortable"

# below the other application.load statements:
application.load(sortableControllers)
```

Note, this PR removes the old `sortable.js` so you'll have to update all instances of `data-reorder` as in Step 3 above.

Checklist:

- [x] make sure dragula plugin instance is operational
- [x] provide instructions for switching to Stimulus method
- [x] provide action to re-apply dragula instance to table contents
